_posts/2017-04-04-what-every-developer-should-know-about-ir.md

```diff
@@ -0,0 +1,250 @@
+
+TF * IDF weight can be calculated in most simple form for term T in document D belonging to collection C as: T occurrences in D * (1 / T occurrences in C). Usually frequencies are further normalized logarithmically, as we probably want to raise weight of those documents with one or two mentions and diminish the juice on those documents which just repeat the same term over and over, but let's keep things simple.
+
+| id    |  doc                                                                  |
+|-------|-----------------------------------------------------------------------|
+| 1     | A dog eats canned dog food. A Dog do not eat chocolate.               |
+| 2     | Cats eat mice.                                                        |
+| 3     | Cats are usually cute. Dogs too.                                      |
+
+TF for couple of terms in documents.
+
+| term  | TF in documents                            | TF in collection        |
+|-------|----------------------------------------------------------------------|
+| dog   | **1:** 3 **2:** 0 **3:** 1 **4:** 0        | 4                       |
+| cat   | **1:** 0 **2:** 1 **3:** 1 **4:** 0        | 3                       |
+
+By the formula we can calculate that weight for "dog" is 0.75 `= 3 * (1 / 4)` for document 1 and 0.25 `= 1 * (1 / 4)` for document 2. From this we reason that document 1 is more relevant than documents 2 for term "dog" and thus should be raised higher in resultset. Note that weight is not necessarily used to reason if the document is found at all but to order the resultset.
+
+Matching algorithms is a vast subject, of which entire books are written, so it can't be thoroughly addressed here. Suggested further reading can be found at the end of this post.
+
+## Vector space vs. boolean model
+
+These models are usually kept clearly separate in research and literature. Most of the real life information retrieval systems are actually hybrid of these using ideas from various sources. For instance predominant Elasticsearch is a hybrid of boolean model and vector space model interpretation. It uses boolean model for matching documents, but utilizes vector space influenced sorting algorithm to sort result set. Postgres has a really good full text search support and it utilizes vector space model pretty cleanly.
+
+Google was originally really simple TF * IDF vector space model system with "page rank" added. Page rank added more weight to pages which are broadly linked from other pages and simirlarly diminished weight of the pages that were seldom linked. For now, we don't know basically anything how Google actually works.
```

Comment:
Google was -> Google had? Google utilized?
_The_ page rank 
from _the_ other pages
and simi-r-larly